egl: add support for libglvnd

peat-psuwit opened the pull request:
libglvnd is "the GL Vendor-Neutral Dispatch library". It supports
arbitrating calls to multiple vendor libraries at runtime, and replaces
the old ways of selecting vendor libraries in modern distros.

This PR consists of a few general/improvement commits that are not directly GLVND-related, followed by the GLVND support itself in a single commit. It does not make sense to split the GLVND support further.

This will be tested in the Focal port of Ubuntu Touch. The PR is opened early to receive feedbacks from maintainers.

krnlyng commented:
There seem to be some quite nice fixes in this, what kind of feed back do you need?

peat-psuwit commented:
On the general approach and such (and if upstream wants this at all). I've successfully run a simple Qt application on it, but haven't run more complex applications. My intent is to un-draft it when I can run Lomiri with this, as the complexity should allows it to hit edge cases.

mlehtima commented:
I saw a new build warning when doing test build of this:
```
egl.c: In function '_my_eglCreatePlatformWindowSurfaceEXT':
egl.c:330:45: warning: passing argument 3 of 'eglCreateWindowSurface' makes integer from pointer without a cast [-Wint-conversion]
   return eglCreateWindowSurface(dpy, config, native_window, attrib_list);
                                                                          ^~~~~~~~~~~~~
egl.c:295:23: note: expected 'EGLNativeWindowType' {aka 'long unsigned int'} but argument is of type 'void *'
    EGLNativeWindowType win,
```

mlehtima commented on hybris/egl/ws.c:
```diff
@@ -27,117 +27,135 @@
+	pthread_mutex_lock(&mutex);
+	if (ws != NULL) {
 		pthread_mutex_unlock(&mutex);
+		return ensureCorrectWs(egl_platform);;
```
Style nitpick, two ; at the end.

peat-psuwit replied:
Fixed.

mlehtima commented on hybris/egl/ws.c:
```diff
@@ -27,117 +27,135 @@
+	const char *eglplatform_dir = PKGLIBDIR;
+	const char *user_eglplatform_dir = getauxval(AT_SECURE)
+										? NULL
+										: getenv("HYBRIS_EGLPLATFORM_DIR");
```
Style nitpick, the indentation of these lines doesn't look good now, use the original formatting or less indentation.

peat-psuwit replied:
Opps. It happens because I set my editor to render a tab character as 4 spaces. That's fixed now.

mlehtima reviewed (commented):
Is there some reason not to squash the last two commits which fix issues introduced in earlier commits in to the commits those are fixing?

peat-psuwit commented:
> Is there some reason not to squash the last two commits which fix issues introduced in earlier commits in to the commits those are fixing?

Indeed there isn't (can't remember why I didn't). So, they're squashed now.

Changes since previous version:
- Some comment grammar fixes. A comment correction.
- Add explicit cast in `_my_eglCreatePlatformWindowSurfaceEXT()` to silence the warning.
- Remove an extra whitespace in `__eglGLVNDGetVendorString()`.
- Styling fixes.

saukko commented on hybris/egl/glvnd/generate/eglFunctionList.py:
```diff
@@ -0,0 +1,238 @@
+"""
```
Missing license header is this also under MIT like other added .py files?

peat-psuwit replied:
Probably yes; the file originally comes from Mesa. The original file doesn't have the notice either, but I think we can assume the same license as the rest of Mesa.
https://github.com/mesa3d/mesa/blob/main/src/egl/generate/eglFunctionList.py

Although, I would say that this file contains mostly data and not much code, and so might be not copyrightable. In any case, I'll update another version with the header some time soon.

saukko commented:
Just a generic note that we should add some description to the git root that states what parts are under what license, at least currently Apache 2.0 and MIT license in the git tree, didn't check if something else as well. Not mantory to fix in PR but just a note :)

mlehtima commented on hybris/egl/egl.c:
```diff
@@ -139,9 +171,69 @@ struct _EGLDisplay *hybris_egl_display_get_mapping(EGLDisplay display)
+
+		case EGL_PLATFORM_ANDROID_KHR:
+			// "null" ws passthrough everything, which essentially means
+			// the Android platform. Beware confustion with NULL value.
```
Typo: confustion should be confusion.

peat-psuwit replied:
I've reworded it completely. There shouldn't be a typo this time, because the phrase is copied from Wikipedia's template. :D

mlehtima commented on hybris/egl/glvnd/eglglvnd.cpp:
```diff
@@ -0,0 +1,149 @@
+{
+    if (name == __EGL_VENDOR_STRING_PLATFORM_EXTENSIONS) {
+        return
+             "EGL_KHR_platform_android"
```
Formatting: indentation doesn't match the lines below.

peat-psuwit replied:
I've aligned the letter `E`, not the quotation mark, with the second next line. The idea was that we need a space for each token, but I don't want an additional space at the start of the string.

Anyway, have a look at Mesa's code, and they align the quotation marks. So I followed you and Mesa.

mlehtima commented on hybris/egl/glvnd/eglglvnd.cpp:
```diff
@@ -0,0 +1,149 @@
+static void *
+__eglGLVNDGetProcAddress(const char *procName)
+{
+    if (strcmp(procName ,"eglQueryString") == 0)
```
Formatting: space should be after "," not before.

peat-psuwit replied:
Fixed.

mlehtima commented on hybris/egl/glvnd/eglglvnd.cpp:
```diff
@@ -0,0 +1,149 @@
+{
+    if (EGL_VENDOR_ABI_GET_MAJOR_VERSION(version) !=
+        EGL_VENDOR_ABI_MAJOR_VERSION)
+       return EGL_FALSE;
```
Indentation is multiples of 4 in other parts of the file. For readability maybe indent the second line in if condition with another 4.

peat-psuwit replied:
Fixed.

mlehtima commented on hybris/egl/ws.c:
```diff
@@ -27,117 +27,135 @@
+	assert(ws != NULL);
+	ws->init_module(&hybris_egl_interface);
+
+	// Keep the name of choosen ws so that future calls can check.
```
Typo: choosen should be chosen.

peat-psuwit replied:
Fixed.

mlehtima reviewed (commented):
Some small typos and style issues found in the code. Otherwise looks good. Tested the non-glvnd parts and seemed to work correctly on a device.

bigOconstant commented:
Any progress on getting this merged? I'm looking forward to having some hardware acceleration on my oneplus 6t

krnlyng commented:
> Any progress on getting this merged? I'm looking forward to having some hardware acceleration on my oneplus 6t

A few things still need to be resolved (see above), but getting close.

peat-psuwit commented:
Sorry for the delay. I forgot about this change for a while, until @krnlyng's comment pinged me.

Other than responding to the reviews:
- Update the EGL header update commit. By the time this is merged, it's probably not-so-latest anymore.
- Update the override of `eglDestroyImageKHR()`, as per the recent commit.

So, please take another look.

MrCyjaneK commented:
I'm getting `undefined symbol: eglCreatePlatformWundowSurface` when launching some apps on droidian, will this pull request fix it?

krnlyng commented:
@peat-psuwit i think the (c) headers are still missing for these two

https://github.com/libhybris/libhybris/blob/362efd97cd5b18f580c402d3fc14555f1ef6e31e/hybris/egl/glvnd/egldispatchstubs.c
https://github.com/libhybris/libhybris/blob/362efd97cd5b18f580c402d3fc14555f1ef6e31e/hybris/egl/glvnd/egldispatchstubs.h

otherwise i think i am ready to merge this.

mlehtima commented:
Also these files are missing copyright
https://github.com/libhybris/libhybris/blob/362efd97cd5b18f580c402d3fc14555f1ef6e31e/hybris/egl/glvnd/generate/egl_hybris.xml
https://github.com/libhybris/libhybris/blob/362efd97cd5b18f580c402d3fc14555f1ef6e31e/hybris/egl/glvnd/generate/egl_other.xml

mlehtima commented:
@peat-psuwit Do you have plans to add the missing copyright headers mentioned in above comments so we could get this merged?

peat-psuwit commented:
Sorry, forgot about this again. Copyright headers are added in the latest commit. Please take another look.

krnlyng commented:
thanks!

peat-psuwit commented:
Wait a minute, why did I introduce an "advertising clause" in egldispatchstub.*?

peat-psuwit commented:
Turns out, the advertising clause might cause some problem...

https://gitlab.freedesktop.org/glvnd/libglvnd/-/issues/221

krnlyng commented:
@peat-psuwit seems glvnd did fix the license terms. do we need to fix something here as well?
